File: gateway/safebox/lora.py

import config
from Crypto.Cipher import AES
import crypto
import PyLora
import struct
import time
import numbers
import os

iv = b'1212121212121212'

def close():
    PyLora.close()

def handle(packet):
    
    if(packet == None): return packet
    op = packet[0]
    size = len(packet)

    if op == config.OP_ACK:
        if size != config.OP_ACK_SIZE:
            raise Exception('invalid ACK packet size {}'.format(size))
    elif op == config.OP_CONNECT:
        if size != config.OP_CONNECT_SIZE:
            raise Exception('invalid CONNECT packet size {}'.format(size))
    elif op == config.OP_STATUS:
        if size != config.OP_STATUS_SIZE:
            raise Exception('invalid STATUS packet size {}'.format(size))
    else:
        raise Exception('invalid OP packet {}'.format(op))

    return packet

def init():
    PyLora.set_pins(cs_pin=config.LORA_SS_PIN, rst_pin=config.LORA_RST_PIN, irq_pin=config.LORA_DIO0_PIN)
    PyLora.init()
    PyLora.set_spreading_factor(config.LORA_SPREADING)
    PyLora.set_preamble_length(config.LORA_PREAMBLE)
    PyLora.set_frequency(config.LORA_FREQUENCY)
    PyLora.set_sync_word(config.LORA_SYNC_WORD)

def packet_str(packet):
    s = ''
    for b in packet:
        s += str(b)
    return s

def packet_check(packet):
    try:
        """
        packetCheck = bytearray(config.OP_STATUS_SIZE)
        packetCheck[0] = config.OP_STATUS
        packetCheck[1:9] = crypto.get_checksum(packetCheck)
        print(str(packetCheck[1:9])+' : '+str(packet[1:9]))
        if packet[1:9] == packetCheck[1:9]:
            return True
        return False
        """
        
        op = str(packet[0]).encode('hex')
        #op = bytes(packet[0]).encode('hex')
        size = 0
        print('op: ' + str(op))
        #print('opInt: ' + int(op))
        print('ACK: ' + str(config.OP_ACK))
        print('ACKHex: ' + str(config.OP_ACK).encode('hex'))

        if op == '0' + str(config.OP_ACK):
        #if op == bytes(config.OP_ACK):
            size = config.OP_ACK_SIZE
        elif op == '0' + str(config.OP_CONNECT):
        #elif op == bytes(config.OP_CONNECT):
            size = config.OP_CONNECT_SIZE
        elif op == '0' + str(config.OP_STATUS):
        #elif op == bytes(config.OP_STATUS):
            size = config.OP_STATUS_SIZE
        else:
            size = 0
        
        print ('Size: ' + str(size))
        
        if size != 0:
            """
            packetCheck = bytearray(size)
            i = 0
            while i < size:
                if i < 9 and i > 0:
                    i+=1
                else:
                    packetCheck[i] = packet[i]
                    i+=1
            
            check = crypto.get_checksum(packetCheck)
            print('Packet: ' + str(packet) + ' Check: ' + str(packetCheck) + ' Checksum: ' + str(check))
            
            if packet[1:9] == check:
                return True
            else:
                return False
            """
            return True
        else:
            return False     
    except:
        return False
        
        
def decrypt_packet(packet):
    i = 0
    for x in os.environ['GATE_PUB_KEYS'].split(','):
        try:
            print(x+' : '+x[16:32])
            aes = AES.new(x[16:32], AES.MODE_CBC, iv)
            decp = aes.decrypt(packet)
            print('ClearText: ' + str(decp).encode('hex'))
            check = packet_check(decp)
            if check:
                return decp, i
        except:
            print(x+' : '+x[16:32])
        i += 1
    return None, -1
    

def recv():
    PyLora.receive()
    if not PyLora.packet_available():
        return None, None

    packet = PyLora.receive_packet()
    print ('Packet: ' + str(packet).encode('hex'))
    decp, boxNum = decrypt_packet(packet)                
    
    #return handle(decp), boxNum
    return decp, boxNum

def recv_wait():
    PyLora.receive()
    while not PyLora.packet_available():
        time.sleep(0)

    packet = PyLora.receive_packet()
    return handle(packet)

def check_topic(msg):
    i = 0
    for x in os.environ['KIT_CHANNEL'].split(','):
        if 'channels/'+x+'/messages' == msg.get_topic():
            return i  
        i+=1
    return -1

def send(packet, boxNum):
    PyLora.idle()
    time.sleep(1)
    #pad and encrypt with box[0:16]
    print('Packet: ' + str(packet).encode('hex') + ' Packet length: ' + str(len(packet)) + ' Box: ' + str(boxNum))
    """
    i = len(packet)
    maxLen = 16
    while i > maxLen:
        maxLen *= 2
    while i < maxLen:
        #packet[i] = 0
        packet.append('0')
        i+=1
    """
    padding = get_padding(len(packet))
    pad_packet = bytearray(len(packet)+padding)
    i=0
    while i < len(packet):
        pad_packet[i] = packet[i]
        i+=1
    print('Packet: ' + str(pad_packet).encode('hex') + ' Packet length: ' + str(len(pad_packet)))
    key = os.environ['GATE_PUB_KEYS'].split(',')[boxNum]
    print ('Key: ' + key)
    aes = AES.new(key[:16], AES.MODE_CBC, iv)
    encd = aes.encrypt(pad_packet)
    print('Cipher: ' + encd.encode('hex') + ' ' + str(len(encd)))
    #decp, boxNum = decrypt_packet(encd)
    #print('ClearText: ' + decp.encode('hex') + ' Box: ' + str(boxNum))
    PyLora.send_packet(encd)

def get_padding(size):
    padding = 16-size%16
    if padding == 16:
        padding = 0
    return padding

def send_allow(boxNum, pin):
    padding = get_padding(config.OP_ALLOW_SIZE)
    packet = bytearray(config.OP_ALLOW_SIZE+padding)
    packet[0] = config.OP_ALLOW
    packet[1:9] = crypto.get_checksum(packet)
    packet[13:15] = struct.pack('>H', pin)
    send(packet, boxNum)

def send_deny(boxNum, pin):
    padding = get_padding(config.OP_DENY_SIZE)
    packet = bytearray(config.OP_DENY_SIZE+padding)
    packet[0] = config.OP_DENY
    packet[1:9] = crypto.get_checksum(packet)
    packet[13:15] = struct.pack('>H', pin)
    send(packet, boxNum)

def send_connect(boxNum, msg):
    padding = get_padding(config.OP_CONNECT_SIZE)
    packet = bytearray(config.OP_CONNECT_SIZE+padding)
    packet[0] = config.OP_CONNECT
    packet[5:] = crypto.get_hash(packet)
    send(packet, boxNum)

def send_lock(boxNum):
    padding = get_padding(config.OP_LOCK_SIZE)
    packet = bytearray(config.OP_LOCK_SIZE+padding)
    packet[0] = config.OP_LOCK
    packet[5:] = crypto.get_checksum(packet)
    send(packet, boxNum)

def send_register(boxNum):
    padding = get_padding(config.OP_REGISTER_SIZE)
    packet = bytearray(config.OP_REGISTER_SIZE+padding)
    packet[0] = config.OP_REGISTER
    packet[5:] = crypto.get_checksum(packet)
    send(packet, boxNum)

def send_status(boxNum):
    padding = get_padding(config.OP_STATUS_SIZE)
    packet = bytearray(config.OP_STATUS_SIZE+padding)
    packet[0] = config.OP_STATUS
    packet[1:9] = crypto.get_checksum(packet)
    send(packet, boxNum)

def send_unlock(boxNum):
    padding = get_padding(config.OP_UNLOCK_SIZE)
    packet = bytearray(config.OP_UNLOCK_SIZE+padding)
    packet[0] = config.OP_UNLOCK
    packet[5:] = crypto.get_checksum(packet)
    send(packet, boxNum)

def sync_word(word):
    PyLora.set_sync_word(word)

File: box4/lora.cpp

#include <LoRa.h>

#include "config.h"
#include "mailbox.h"

unsigned char gw_key[HASH_SIZE] = {0};
bool gw_key_loaded = false;
unsigned char key[HASH_SIZE] = {0};
bool key_loaded = false;

//extern byte aes_iv[16];

void print_packet(unsigned char *buffer, int size) {
    for (int i = 0; i < size; i++) {
        Serial.print(buffer[i], HEX);
        Serial.print(" ");
    }
    Serial.println();
}

int lora_init() {
    int error = load_gw_key(gw_key);
    if (!error) {
        gw_key_loaded = true;
        if (DEBUG) {
            Serial.print("LORA: loaded key ");
            print_packet(gw_key, HASH_SIZE);
        }
    }

    error = load_key(key);
    if (!error) {
        key_loaded = true;
        if (DEBUG) {
            Serial.print("LORA: loaded key ");
            print_packet(key, HASH_SIZE);
        }
    } else {
        unsigned char secret[] = SECRET_KEY;
        int size = sizeof(SECRET_KEY);
        if (size > HASH_SIZE) size = HASH_SIZE;
        memcpy(key, secret, size);
        if (DEBUG) {
            Serial.print("LORA: loaded secret ");
            print_packet(key, HASH_SIZE);
        }
    }

    LoRa.setPins(LORA_SS_PIN, LORA_RST_PIN, LORA_DIO0_PIN);
    //LoRa.setSPIFrequency(1e6);
    if (!LoRa.begin(LORA_FREQUENCY)) return E_LORA;

    LoRa.setSpreadingFactor(LORA_SPREADING);
    LoRa.setSyncWord(LORA_SYNC_WORD);
    return E_OK;
}

int lora_handle(const uint8_t *data, int size) {
    uint8_t op = data[0];
    //if (op >= OP_END) return E_LORA_OP_CODE;

    uint16_t pin = 0;
    int result = E_OK;
    switch (op) {
        case OP_ACK:
            //if (size != OP_ACK_SIZE) return E_LORA_MAX_SIZE;
            return E_OK;
        case OP_ALLOW:
            //if (size != OP_ALLOW_SIZE) return E_LORA_MAX_SIZE;
            pin = (data[13] << 8) | data[14];
            result = allow_set(String(pin));
            break;
        case OP_CONNECT:
            //if (size != OP_CONNECT_SIZE) return E_LORA_MAX_SIZE;
            break;
        case OP_DENY:
            //if (size != OP_DENY_SIZE) return E_LORA_MAX_SIZE;
            pin = (data[13] << 8) | data[14];
            result = allow_del(String(pin));
            break;
        case OP_LOCK:
            Serial.println("lock check");
            //if (size != OP_LOCK_SIZE) return E_LORA_MAX_SIZE;
            Serial.println("lock pass");
            result = set_lock(LOCK_YES);
            break;
        case OP_REGISTER:
            //if (size != OP_REGISTER_SIZE) return E_LORA_MAX_SIZE;
            if (DEBUG) Serial.println("LORA: setting sync word");
            LoRa.setSyncWord((int)data[3]);
            break;
        case OP_STATUS:
            //if (size != OP_STATUS_SIZE) return E_LORA_MAX_SIZE;
            break;
        case OP_TIME:
            //if (size != OP_TIME_SIZE) return E_LORA_MAX_SIZE;
            break;
        case OP_UNLOCK:
            //if (size != OP_UNLOCK_SIZE) return E_LORA_MAX_SIZE;
            result = set_lock(LOCK_NO);
            break;
        default:
            return E_LORA_OP_CODE;
    }

    if (op == OP_CONNECT) {
        //if (!gw_key_loaded) {
            if (DEBUG) Serial.println("LORA: saving new gw key");
            result = save_gw_key(gw_key);
            if (result) return result;
        //}

        if (DEBUG) Serial.println("LORA: building CONNECT reply");
        unsigned char buffer[OP_CONNECT_SIZE] = {0};

        buffer[0] = OP_CONNECT;
        buffer[1] = get_flag();
        buffer[2] = get_lock();
        buffer[3] = get_package();
        buffer[4] = get_power();

        // SECRET_KEY will be used as the initial hash value
        // with remaining checksums using the hash of it.
        if (!key_loaded) get_hash(key, buffer, OP_CONNECT_SIZE, key);
        for (int i = 0; i < HASH_SIZE; i++) {
            buffer[i + 5] = key[i];
        }

        if (!key_loaded) {
            result = save_key(key);
            if (result) return result;
        }

        if (DEBUG) print_packet(buffer, OP_CONNECT_SIZE);
        result = lora_send(buffer, OP_CONNECT_SIZE);
    } else {
        if (DEBUG) Serial.println("LORA: building ACK reply");
        unsigned char buffer[OP_ACK_SIZE] = {0};

        buffer[0] = OP_ACK;

        unsigned char checksum[CHECKSUM_SIZE];
        get_checksum(key, buffer, OP_ACK_SIZE, checksum);
        for (int i = 0; i < CHECKSUM_SIZE; i++) {
            buffer[i + 1] = checksum[i];
        }

        buffer[9] = get_flag();
        buffer[10] = get_lock();
        buffer[11] = get_package();
        buffer[12] = get_power();
        buffer[13] = (uint8_t)result;

        if (DEBUG) print_packet(buffer, OP_ACK_SIZE);
        result = lora_send(buffer, OP_ACK_SIZE);
    }

    return result;
}

int lora_recv() {
    int size = LoRa.parsePacket();
    if (!size) return E_OK;
    if (DEBUG) {
        Serial.println();
        Serial.print("LORA: received packet ");
        Serial.println(size, DEC);
        Serial.println("LORA: RSSI ");
        Serial.println(LoRa.packetRssi());
    }
    //if (size > LORA_MAX_SIZE) return E_LORA_MAX_SIZE;

    Serial.println("recv check");
    uint8_t data[size] = {0};
    for (int i = 0; LoRa.available() && i < size; i++) {
        data[i] = (uint8_t)LoRa.read();
    }

    Serial.print("Data: "); Serial.write((uint8_t *)data, sizeof(data)); Serial.println("");
    print((char*)data,size);

    /*
    char msg[size];
    for (int i = 0; i<size; i++) {
      msg[i] = (char)data[i];
    }
    
    char msg[BLOCK_SIZE];
    memset(msg, 0, BLOCK_SIZE);
    sprintf(msg, "%s", data);
    */
    //print(msg, size);
    /*
    uint8_t msg_data[] = {0x07, 0xc6, 0xa7, 0x85, 0x80, 0x38, 0x40, 0xee, 0x28, 0x01, 0x01, 0x01, 0x01, 0x00, 0x00, 0x00};
    char testmsg[BLOCK_SIZE];
    memset(testmsg, 0, BLOCK_SIZE);
    sprintf(testmsg, "%s", msg_data);
    Serial.print("MESSAGE: "); Serial.println(testmsg);
    
    unsigned char cipher[BLOCK_SIZE];
    encrypt(cipher, testmsg, BLOCK_SIZE);
    Serial.print("ENCRYPT: "); print((char *)cipher, BLOCK_SIZE);
    */
    
    unsigned char cleartxt[size];
    unsigned char iv[] = "1212121212121212";
    decrypt(cleartxt,(char*) data, size);
    Serial.print("DECRYPT: "); Serial.write((uint8_t *)cleartxt, BLOCK_SIZE); Serial.println("");
    print((char*)cleartxt, sizeof(cleartxt));

    if (DEBUG) print_packet(cleartxt, sizeof(cleartxt));
    int result = lora_handle(cleartxt, sizeof(cleartxt));
    return result;
}

int get_padding(int size) {
    int padding = 16-size%16;
    if(padding == 16){
        padding = 0;
    }
    return padding;
}
int lora_send(unsigned char *buffer, int size) {
    if (!LoRa.beginPacket()) return E_LORA_BEGIN_PACKET;
    
    int padding = get_padding(size) + size;
    char msg[padding];
    memset(msg, 0, padding);
    memcpy(msg, buffer, size);
    Serial.print("Message: "); Serial.println((char*)buffer);
    Serial.print("MessagePad: "); Serial.println(msg);
    Serial.print("MessagePadHex: "); print(msg,padding);
    Serial.print("Padding: "); Serial.print(padding, DEC); Serial.print(" Size: "); Serial.println(size, DEC);
    Serial.print("MessageSize: "); Serial.println(sizeof(msg), DEC);

    
    //encrypt buffer with key[15:]
    unsigned char cipher[padding];
    //unsigned char iv[] = "1212121212121212";
    encrypt(cipher, msg, padding);

    Serial.print("Cipher: "); print((char*)cipher, padding);
    
    size_t size_wrote = LoRa.write((uint8_t *)cipher, padding);
    Serial.print("Lora Write: "); Serial.println(size_wrote);
    if (padding != size_wrote) return E_LORA_SIZE_WROTE;

    LoRa.endPacket();
    //if (!LoRa.endPacket()) return E_LORA_END_PACKET;
    return E_OK;
}
